xdsl/transforms/accfg/trace_acc_state.py

```diff
@@ -0,0 +1,74 @@
+"""
+These inference helpers are used to figure out which values are
+present at which points in the IR.
+
+Example inference results:
+
+```
+accfg.setup(A = %1, B = %2)  // previous state = {}
+// ...
+accfg.setup(A = %3, C = %1)  // previous state = {A = %1, B = %2}
+```
+
+These inference passes walk the IR backwards.
+"""
+
+from xdsl.dialects import accfg, scf
+from xdsl.ir import Block, SSAValue
+
+State = dict[str, SSAValue]
+
+
+def infer_state_of(state_var: SSAValue) -> State:
```

Comment:
Would it not be useful to stop early? If the same values keep getting updated, it seems like we'd want to stop after seeing the latest definitions, not go up the whole chain every time.

Reply:
It looks like the complexity of getting the known states for n ops is O(n^2) right now, if I understand correctly

Reply:
@superlopuh, That's a very good comment. I think the complexity gets even worse (exponentially) the more if statements you have. We can probably stop early if we carry the currently known state with the recursive function and only update the values that are not known. However, right now that means I have to do a symbol table lookup on the accelerator symbol that specifies the amount of parameters that have to be set to get a stopping criterion. Is there a canonical way of accessing the symbol table if the current scope is only the op in that module?

Reply:
I would add optimizations when we need to, for now, I'd keep the code as simple as possible.

If we decide to change something, it's infinitely easier to edit un-optimized code.

Reply:
My point is that I'm not sure that this is as simple as possible :) @JosseVanDelm, what do you mean by "canonical way of accessing the symbol table?"

Reply:
I think I was looking for:
```
SymbolTable.lookup_symbol
```

Reply:
Ah yeah that was my first thought but I didn't realise that it's what you were looking for :)

Reply:
If okay I'd also like to postponing this as an optimization. I think we have other more pressing issues to work on first :D